### PR TITLE
fixed channel state clearing

### PR DESCRIPTION
**Pull request by hackaugusto**
the channel state must be cleared only after all unlocks have been done.
After the close/updateTransfer/settle cycle, there are three
scenarios:

- None of the participants had pending locks, so both merkle trees are
  empty and the channel can be cleared at settle time
- One of the participants had pending locks, this may be the current
  node or its partner. Under this case the channel should be cleared
  after the only unlock.
- Both partcipants had pending locks. In this case the channel should be
  cleared after both unlocks.

To implement the above this is changing the `onchain_locksroot`
attribute during ContractReceiveChannelUnlock, and if both locksroots
are EMPTY then the channel can be safely cleared.

fixes: #4024

### CODE REVIEW

File: raiden/tests/unit/test_channelstate.py [hunk 1]
```diff
@@ -95,15 +105,15 @@ def create_model(balance, merkletree_width=0):
         amount_locked=0,
         balance=balance,
         distributable=balance,
-        next_nonce=1,
+        next_nonce=len(merkletree_leaves) + 1,
```
Comment (hackaugusto) [inline raiden/tests/unit/test_channelstate.py]: This is the lowest possible value for the balance proof nonce, since we need at least one `LockedTransfer` message per lock.

File: raiden/tests/unit/test_channelstate.py [hunk 2]
```diff
@@ -123,6 +133,43 @@ def create_channel_from_models(our_model, partner_model):
...
+        )
+
+        partner_signed = make_signed_balance_proof_from_unsigned(
+            partner_unsigned, LocalSigner(partner_pkey)
+        )
+    else:
+        partner_signed = None
+
+    channel_state.our_state.balance_proof = our_unsigned
+    channel_state.partner_state.balance_proof = partner_signed
```
Comment (hackaugusto) [inline raiden/tests/unit/test_channelstate.py]: The factory was missing the balance proofs. I had to add this because clearing the channel state depends on the locksroot, and the value for the BP is used.

File: raiden/tests/unit/test_channelstate.py [hunk 3]
```diff
@@ -95,15 +105,15 @@ def create_model(balance, merkletree_width=0):
...
+        next_nonce=len(merkletree_leaves) + 1,
         merkletree_leaves=merkletree_leaves,
         contract_balance=balance,
     )
 
     return our_model, privkey
 
 
-def create_channel_from_models(our_model, partner_model):
+def create_channel_from_models(our_model, partner_model, partner_pkey):
```
Comment (hackaugusto) [inline raiden/tests/unit/test_channelstate.py]: because the factory now creates valid balance proofs we need the partner private key to create the signed balance proof

File: raiden/transfer/channel.py [hunk 4]
```diff
@@ -1790,17 +1790,28 @@ def apply_channel_newbalance(
...
+        partner_state = channel_state.partner_state
+        if state_change.participant == our_state.address:
+            our_state.onchain_locksroot = EMPTY_MERKLE_ROOT
+        elif state_change.participant == partner_state.address:
+            partner_state.onchain_locksroot = EMPTY_MERKLE_ROOT
+
+        # only clear the channel state once all unlocks have been done
+        no_unlock_left_to_do = (
+            our_state.onchain_locksroot == EMPTY_MERKLE_ROOT
+            and partner_state.onchain_locksroot == EMPTY_MERKLE_ROOT
```
Comment (hackaugusto) [inline raiden/transfer/channel.py]: This is the fix, very simple, we just wait for the necessary `ContractReceiveChannelBatchUnlock`. It can be either one or two unlocks.

File: raiden/tests/unit/test_tokennetwork.py [hunk 5]
```diff
@@ -208,15 +209,18 @@ def test_channel_data_removed_after_unlock(
...
 
     settle_block_number = closed_block_number + channel_state.settle_timeout + 1
     channel_settled_state_change = ContractReceiveChannelSettled(
         transaction_hash=factories.make_transaction_hash(),
         canonical_identifier=channel_state.canonical_identifier,
         block_number=settle_block_number,
         block_hash=factories.make_block_hash(),
-        our_onchain_locksroot=factories.make_32bytes(),
-        partner_onchain_locksroot=EMPTY_MERKLE_ROOT,
+        our_onchain_locksroot=merkleroot(channel_state_after_closed.our_state.merkletree),
```
Comment (hackaugusto) [inline raiden/tests/unit/test_tokennetwork.py]: now we must use a valid merkleroot value, otherwise the channel cleaning does not work.

File: raiden/tests/unit/test_tokennetwork.py [hunk 6]
```diff
@@ -343,8 +347,8 @@ def test_mediator_clear_pairs_after_batch_unlock(
     channel_batch_unlock_state_change = ContractReceiveChannelBatchUnlock(
         transaction_hash=factories.make_transaction_hash(),
         canonical_identifier=channel_state.canonical_identifier,
-        participant=our_address,
-        partner=address,
+        participant=address,
```
Comment (hackaugusto) [inline raiden/tests/unit/test_tokennetwork.py]: these were swapped around, the unlock happens at the `partner` end

**Comment by rakanalh** [comment]
Reviewing this

File: raiden/tests/integration/long_running/test_settlement.py [hunk 7]
```diff
@@ -382,14 +381,14 @@ def run_test_batch_unlock(
         in token_network.partneraddresses_to_channelidentifiers[alice_app.raiden.address]
     )
 
-    # wait for the node to call batch unlock
+    # Wait for both nodes to call batch unlock
```
Comment (rakanalh) [inline raiden/tests/integration/long_running/test_settlement.py]: Below this, we're only waiting on bob's app to receive the batch unlock but not alice... we should do it twice there.

File: raiden/tests/integration/long_running/test_settlement.py [hunk 8]
```diff
@@ -322,7 +321,7 @@ def run_test_batch_unlock(
         secret=secret,
     )
 
-    gevent.sleep(1)  # wait for the messages to be exchanged
+    secret_request_event.get()  # wait for the messages to be exchanged
```
Comment (rakanalh) [inline raiden/tests/integration/long_running/test_settlement.py]: I promised i would do this in a separate PR a while ago, thanks for doing these changes

File: raiden/tests/integration/long_running/test_settlement.py (hunk 7 above)
Comment (rakanalh) [reply raiden/tests/integration/long_running/test_settlement.py]: Since the asserts after this wait depend on one side of the channel-state, we agreed not to do this.
